halo/evmstaking/evmstaking.go

```diff
@@ -91,73 +98,81 @@ func (p EventProcessor) Prepare(ctx context.Context, blockHash common.Hash) ([]*
 }
 
 func (EventProcessor) Name() string {
-	return AccountName
+	return ModuleName
 }
 
 func (p EventProcessor) Addresses() []common.Address {
 	return []common.Address{p.address}
 }
 
-// Deliver processes a omni deposit log event:
-// - Mint the corresponding amount of $STAKE coins.
-// - Send the minted coins to the depositor's account.
-// - Create a new validator with the depositor's account.ccbelunhfhcrcegtfvbgfibjjllhjurctgfrfjhhh.
+// Deliver processes a omni deposit log event, which must be one of:
+// - CreateValidator
+// - Delegate.
 func (p EventProcessor) Deliver(ctx context.Context, _ common.Hash, elog *evmenginetypes.EVMEvent) error {
-	deposit, err := p.contract.ParseDeposit(elog.ToEthLog())
-	if err != nil {
-		return errors.Wrap(err, "parse deposit")
-	}
+	ethlog := elog.ToEthLog()
 
-	stdPubkey, err := k1util.PubKeyFromBytes64(deposit.Pubkey)
-	if err != nil {
-		return errors.Wrap(err, "deposit pubkey")
+	switch ethlog.Topics[0] {
+	case createValidatorEvent.ID:
+		ev, err := p.contract.ParseCreateValidator(ethlog)
+		if err != nil {
+			return errors.Wrap(err, "parse create validator")
+		}
+
+		return p.deliverCreateValidator(ctx, ev)
+	case delegateEvent.ID:
+		ev, err := p.contract.ParseDelegate(ethlog)
+		if err != nil {
+			return errors.Wrap(err, "parse delegate")
+		}
+
+		return p.delivereDelegate(ctx, ev)
+	default:
+		return errors.New("unknown event")
 	}
+}
 
-	pubkey, err := k1util.StdPubKeyToCosmos(stdPubkey)
+// deliverCreateValidator processes a CreateValidator event, and creates a new validator.
+// - Mint the corresponding amount of $STAKE coins.
+// - Send the minted coins to the depositor's account.
+// - Create a new validator with the depositor's account.
+//
+// NOTE: if we error before coins are sent, the deposit is lost.
+// NOTE: if we error after coins are sent, the account will have $STAKE tokens minted to it, with no way of accessing.
+// TODO: consider recovery methods for the above scenarios.
```

Comment:
These cases worth discussing

Reply:
errors result in rollback of all updated state, so nothing to worry about, any error results in no state updates

Reply:
oh sweet